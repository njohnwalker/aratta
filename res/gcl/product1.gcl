@req(m >= 0 & n >= 0)
m1, n1, res := 0, 0, 0 ;
do @inv()
// invariant == m1 <= m & res == m1 * n
[]  m1 < m ->
    n1 := 0 ;
    do @inv()
    // invariant n1 <= n & res == m1 * n + n1
    [] n1 < n -> res, n1 := res + 1, n1 + 1;
    od
    m1 := m1 + 1 ;
od
@ens(res == m * n)
