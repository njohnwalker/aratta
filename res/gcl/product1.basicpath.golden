Path 1
  ; Substitute
    (m1, 0)
    (n1, 0)
    (res, 0)
  { Satisfies INVARIANT > PLACEHOLDER }
Path 2
  { Satisfies res == m * n }
Path 3
  ; Assume INVARIANT > PLACEHOLDER
  ; Branch
    ; Assume m1 < m
    ; Substitute
      (n1, 0)
    { Satisfies INVARIANT > PLACEHOLDER }
  { Satisfies true }
Path 4
  ; Assume INVARIANT > PLACEHOLDER
  ; Branch
    ; Substitute
      (m1, m1 + 1)
    { Satisfies INVARIANT > PLACEHOLDER }
  { Satisfies true }
Path 5
  ; Assume INVARIANT > PLACEHOLDER
  ; Branch
    ; Assume INVARIANT > PLACEHOLDER
    ; Branch
      ; Assume n1 < n
      ; Substitute
        (res, res + 1)
        (n1, n1 + 1)
      { Satisfies INVARIANT > PLACEHOLDER }
    { Satisfies true }
  { Satisfies true }